File: ds_base/ds_param/src/ds_param_example.cpp
/**
* Copyright 2018 Woods Hole Oceanographic Institution
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are met:
*
* 1. Redistributions of source code must retain the above copyright notice,
*    this list of conditions and the following disclaimer.
*
* 2. Redistributions in binary form must reproduce the above copyright notice,
*    this list of conditions and the following disclaimer in the documentation
*    and/or other materials provided with the distribution.
*
* 3. Neither the name of the copyright holder nor the names of its contributors
*    may be used to endorse or promote products derived from this software
*    without specific prior written permission.
*
* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
* ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
* LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
* CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
* SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
* INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
* CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
* ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
* POSSIBILITY OF SUCH DAMAGE.
*/
//
// Created by ivaughn on 1/30/18.
//

#include "ds_param/ds_param_conn.h"
#include "ds_param/ds_param_guard.h"

#include <boost/optional/optional_io.hpp>

class ParamDemo
{
public:
  ParamDemo(ros::NodeHandle& _nh) : handle(_nh)
  {
    conn = ds_param::ParamConnection::create(handle);

    // Grab a copy of our STATIC parameters
    // Note that this is a standard ROS call
    handle.getParam(ros::this_node::getName() + "/other_node", other_node);
    handle.getParam(ros::this_node::getName() + "/start_idx", start);

    // connect to OUR shared enum
    param_enum = conn->connect<ds_param::EnumParam>(ros::this_node::getName() + "/test_enum_param", true);
    // setup our shared enum
    param_enum->addNamedValue("Option 1", 1);
    param_enum->addNamedValue("non-consecutive Option 7", 7);

    // connect to OUR shared parameters
    param_int = conn->connect<ds_param::IntParam>(ros::this_node::getName() + "/test_int_param", true);
    param_str = conn->connect<ds_param::StringParam>(ros::this_node::getName() + "/test_str_param", true);

    // connect to the OTHER NODE's shared parameters
    other_int = conn->connect<ds_param::IntParam>("/" + other_node + "/test_int_param", false);
    other_str = conn->connect<ds_param::StringParam>("/" + other_node + "/test_str_param", false);

    // setup two parameters we'll update atomically
    param_bool_atomic = conn->connect<ds_param::BoolParam>(ros::this_node::getName() + "/test_atomic_bool", false);
    param_int_atomic = conn->connect<ds_param::IntParam>(ros::this_node::getName() + "/test_atomic_int", false);

    // test prep
    idx = start;
  }

  void _callback(const ros::TimerEvent& evt)
  {
    std::stringstream ss;
    ss << "Idx " << start << " -> " << idx;

    other_int->Set(idx);
    other_str->Set(ss.str());

    std::string old_str;
    if (param_int->GetPrevious())
    {
      old_str = std::to_string(param_int->GetPrevious().get());
    }
    else
    {
      old_str = "NOT SET";
    }

    // these two will update simulatenously
    {
      // this will prevent updates from going out
      ds_param::ParamGuard lock(conn);

      param_bool_atomic->Set(!param_bool_atomic->Get());
      param_int_atomic->Set(idx);
      if (param_int->GetPrevious())
      {
        ROS_ERROR_STREAM("\tResetting PARAM INT: " << param_int->Get() << " to " << param_int->GetPrevious());
        param_int->Set(*(param_int->GetPrevious()));
        ROS_ERROR_STREAM("\tReset PARAM INT: " << param_int->Get());
      }
    }

    ROS_ERROR_STREAM(ros::this_node::getName()

                     << ": "
                     << "MY param: " << param_int->Get() << " --> \"" << param_str->Get() << "\" (was " << old_str
                     << ")"

                     << "    OTHER param: " << other_int->Get() << " --> \"" << other_str->Get() << "\"");
    idx++;
  }

  // Setup the callback demos
  void setupCallback()
  {
    conn->setCallback(boost::bind(&ParamDemo::_change_callback, this, _1));
  }

  void _change_callback(const ds_param::ParamConnection::ParamCollection& params)
  {
    for (auto iter = params.begin(); iter != params.end(); iter++)
    {
      if (*iter == param_int)
      {
        ROS_ERROR_STREAM("\t\t\tPARAM INT Changed from \"" << param_int->GetPrevious() << "\" to \"" << param_int->Get()
                                                           << "\"");
      }
    }
  }

protected:
  ros::NodeHandle& handle;
  ds_param::ParamConnection::Ptr conn;
  int idx;

  // OUR static paramters
  std::string other_node;
  int start;

  // OUR shared parameters
  ds_param::EnumParam::Ptr param_enum;
  ds_param::IntParam::Ptr param_int;
  ds_param::StringParam::Ptr param_str;

  ds_param::BoolParam::Ptr param_bool_atomic;
  ds_param::IntParam::Ptr param_int_atomic;

  // The OTHER NODE's shared parameters
  ds_param::IntParam::Ptr other_int;
  ds_param::StringParam::Ptr other_str;
};

int main(int argc, char* argv[])
{
  ros::init(argc, argv, "test_node_1");
  ros::NodeHandle nh;
  ParamDemo demo(nh);

  ros::Timer timer = nh.createTimer(ros::Duration(3.0), boost::bind(&ParamDemo::_callback, &demo, _1), false);

  bool use_callback;
  if (!nh.getParam(ros::this_node::getName() + "/use_callback", use_callback))
  {
    ROS_FATAL_STREAM("No variable for use_callback!");
    return -1;
  }

  if (use_callback)
  {
    demo.setupCallback();
  }

  // spins until shut down
  while (ros::ok())
  {
    ros::spin();
  }

  return 0;
}

File: ds_base/ds_param/src_py/ds_param/ds_param_example.py
#!/usr/bin/env python
# Copyright 2018 Woods Hole Oceanographic Institution
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice,
#    this list of conditions and the following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions and the following disclaimer in the documentation
#    and/or other materials provided with the distribution.
#
# 3. Neither the name of the copyright holder nor the names of its contributors
#    may be used to endorse or promote products derived from this software
#    without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
import sys

import rospy
from ds_param import *


class ParamDemo(object):
    def __init__(self):
        # Load some initial parameters
        self.other_node = rospy.get_param(rospy.get_name() + '/other_node')
        self.start_idx  = rospy.get_param(rospy.get_name() + '/start_idx')

        self.conn = ParamConnection()

        # Connect to OUR shared parameters
        self.param_enum = self.conn.connect(rospy.get_name() + '/test_enum_param', EnumParam, True)
        self.param_enum.add_named_value("Option 1", 1)
        self.param_enum.add_named_value("non-consecutive option 6", 6)
        self.param_int = self.conn.connect(rospy.get_name() + '/test_int_param', IntParam, True)
        self.param_str = self.conn.connect(rospy.get_name() + '/test_str_param', StringParam, True)

        # Connect to the OTHER NODE's shared parameters
        self.other_int = self.conn.connect('/' + self.other_node + "/test_int_param", IntParam, False)
        self.other_str = self.conn.connect('/' + self.other_node + "/test_str_param", StringParam, False)

        # Setup two parameters that update together
        self.param_bool_atomic = self.conn.connect(rospy.get_name() + "/test_atomic_bool", BoolParam, False)
        self.param_int_atomic  = self.conn.connect(rospy.get_name() + "/test_atomic_int",   IntParam, False)

        self.index = self.start_idx

    def _timer_callback(self, evt):
        str_to_send = 'Idx %d -> %d' % (self.start_idx, self.index)

        # Send our parameters
        self.other_int.set(self.index)
        self.other_str.set(str_to_send)

        # Read our own
        if self.param_int.get_previous is None:
            old_str = 'NOT SET'
        else:
            old_str = str(self.param_int.get())

        # Update some stuff atomically
        self.conn.lock()

        self.param_bool_atomic.set(not self.param_bool_atomic.get())
        self.param_int_atomic.set(self.index)

        self.conn.unlock()

        rospy.logerr("%s: MY param %d --> \"%s\" (was %s)    OTHER param: %d --> \"%s\"",
                     rospy.get_name(),
                     self.param_int.get(), self.param_str.get(), old_str,
                     self.other_int.get(), self.other_str.get())

        self.index += 1

    def setup_callback(self):
        self.conn.set_callback(self.change_callback)


    def change_callback(self, params):
        rospy.logerr('/------------------------------------------')
        for p in params:
            rospy.logerr("|           PARAM changed from \"%s\" to \"%s\"", p.name(), p.get())
        rospy.logerr('\------------------------------------------')


def main():

    rospy.init_node('ds_param_demo', sys.argv)

    demo = ParamDemo()

    timer = rospy.timer.Timer(rospy.Duration(3.0), demo._timer_callback, oneshot=False)

    use_callback = rospy.get_param(rospy.get_name() + "/use_callback")
    if use_callback:
        demo.setup_callback()

    # Spin until ctrl-C
    rospy.spin()


if __name__ == '__main__':

    main()
